File: leetcode/solutions/0109.ConvertSortedListToBinarySearchTree/sortedListToBST.go
package solution

import (
	"github.com/yukisato/leetcode-util/datastructure"
)

// ListNode structure for LeetCode
type ListNode = datastructure.ListNode

// TreeNode structure for LeetCode
type TreeNode = datastructure.TreeNode

func sortedListToBST(head *ListNode) *TreeNode {
	var nums []int

	for head != nil {
		nums = append(nums, head.Val)
		head = head.Next
	}

	return buildBst(nums)
}

func buildBst(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}

	mid := len(nums) / 2
	node := &TreeNode{Val: nums[mid]}
	node.Left = buildBst(nums[:mid])
	node.Right = buildBst(nums[mid+1:])

	return node
}

File: leetcode/solutions/0109.ConvertSortedListToBinarySearchTree/sortedListToBST.ts
import { BinaryTreeNode } from "../../../datastructure/BinaryTreeNode.ts";
import { SinglyLinkedListNode } from "../../../datastructure/SinglyLinkedListNode.ts";

export default function sortedListToBST(
  head: SinglyLinkedListNode<number> | null
): BinaryTreeNode<number> | null {
  if (!head) {
    return null;
  }

  const nums: number[] = [];

  while (head) {
    nums.push(head.val);
    head = head.next;
  }

  return buildBst(nums, 0, nums.length - 1);
}

function buildBst(
  nums: number[],
  first: number,
  last: number
): BinaryTreeNode<number> | null {
  if (first > last) {
    return null;
  }

  const mid = (first + last) >> 1;
  const node: BinaryTreeNode<number> = {
    val: nums[mid],
    left: null,
    right: null,
  };
  node.left = buildBst(nums, first, mid - 1);
  node.right = buildBst(nums, mid + 1, last);

  return node;
}
